src/main.rs

```diff
@@ -85,33 +58,80 @@ async fn run_single_slot(
     Ok(())
 }
 
+async fn run_oracle_node(
+    price_provider: Arc<impl PriceProvider>,
+    message_generator: MessageGenerator,
+    message_broadcaster: Arc<impl MessageBroadcaster>,
+    slot_provider: Arc<SystemClockSlotProvider>,
+    rx: mpsc::Receiver<()>,
+) -> Result<()> {
+    ReceiverStream::new(rx)
```

Comment:
I moved the "work scheduling" code out of slotsprovider and into main. I don't think main is a great place for it, at the same time I really like how SlotsProviders now only have the responsibility to provide slots and doesn't have to worry about what to do if a handler isn't fast enough or something.

Mmm, that makes me wonder. We could also have it so a slots provider is completely pull based. Ask it any time for the current slot. Have a loop here on 100ms or so, checks what the current slot is, what the last processed slot is, and if the two are apart and we're under the concurrency limit, it starts a new slot run. Could require way fewer lines perhaps and loses any leap seconds issues.

Reply:
Yea I was thinking that myself this morning. Maybe we can start by calling it AttestationScheduler. It can slurp this code out of main. Maybe even absorb most of the SlotsProvider. I might've taken the name to literally.